src/runtime/runtime.cs

```diff
@@ -100,7 +100,7 @@ internal static Version PyVersion
         /// </summary>
         /// <remarks>Always call this method from the Main thread.  After the
         /// first call to this method, the main thread has acquired the GIL.</remarks>
-        internal static void Initialize(bool initSigs = false, ShutdownMode mode = ShutdownMode.Default)
+        internal static void Initialize(bool initSigs = false, ShutdownMode mode = ShutdownMode.Default, bool ensureGIL = true)
```

Comment:
Why do we have `ensureGIL` parameter. Can't caller do it on their own?

Reply:
The problem is the `Ensure` in `Runtime.Initialize` that was added for the different shutdown modes. It doesn't store the new gilstate anywhere which leads to a deadlock once we return (cffi releases the GIL, so we have to take it explicitly). I'm open to suggestions here, but this works for now :)